### PR TITLE
Separate state for tracking downstream end_stream in filter manager and HCM

### LINKED ISSUE #35814: Deferred requests with body or trailers can be proxied incomplete
There are two bugs in processing of deferred requests, that can cause the request to proxied incomplete. I.e request with body is proxied without the body or with partial body. Or request with trailers is proxies without trailers. This bug also requires a decoder filter that pauses and resumes filter chain iteration.

Preconditions:

1. Envoy receives multiple complete requests with body and/or trailers in one I/O read operation.
2. Some of these requests are deferred due to limit on the number of requests processed in one I/O iteration.
3. A decoder filter which pauses and then continues filter chain iteration (i.e. ext_proc).

For example:

1. gRPC client sends multiple small requests, which Envoy reads in one I/O operation.
2. The limit on requests per I/O operation is 1 (or lower than the number sent by the client).
3. ext_proc filter present in the filter chain.
4. The deferred gRPC requests will be proxied without bodies.

Specifically:

1. Filter manager (FM) uses presence of the timestamp of receipt of the downstream end_stream (set when HCM receives end_stream from the client) to set the end_stream during filter chain resumption. If a complete request (with end_stream= true) was deferred, its filter chain iteration would start with the downstream end_stream flag set (since the timestamp was set before the request was deferred). If there is a decoder filter in the chain that pauses and resumes filer chain iteration, all filters after that will see end_stream as true.

For example if a filter pauses in decodeHeaders() and then resumes in decodeData() all filters after that will have decodeHeaders called with end_stream == true. This can cause request to be proxied by the router without the body.

5. The request trailers were also immediately available to the filter manager when processing of deferred request begins. If there is a filter that pauses and resumes iteration it makes filter manager incorrectly process trailers as though they were added by the filter that paused iteration. This causes trailers to be processed by filter chain twice triggering asserts.

The same problem is also present when HCM recreates the stream with body (i.e. for internal redirect).

All presently supported versions v1.28.0 to v1.31.0 are affected.

### PR DESCRIPTION
**Pull request by yanavlasov**
Commit Message:
Fix two bugs in processing of deferred requests.

1. Filter manager (FM) uses presence of the timestamp of receipt of the downstream end_stream to set the end_stream during filter chain resumption. If a complete request (with end_stream= true) was deferred, its filter chain iteration would start with the downstream end_stream flag set. If there is a decoder filter that pauses and resumes filer chain iteration, all filters after resumption will see end_stream as true.

For example if a filter pauses in decodeHeaders() and then resumes in decodeData() all filters after that will have decodeHeaders() called with end_stream == true. This can cause request to be proxied by the router without the body.

2. The request trailers were also immediately available to filter manager when processing of a deferred request begins. If there is a filter that pauses and resumes iteration it makes filter manager incorrectly process trailers as though they were added by the filter that paused iteration. This causes trailers to be processed by filter chain twice triggering asserts.

The same problem is also present when HCM recreates the stream with body (i.e. for internal redirect).

This change makes filter manager use its own, already existing, flag `observed_decode_end_stream_` for tracking state of downstream end_stream flag when resuming the filter chain, instead of the timestamp of the end_stream receipt from downstream client.

HCM continues to use the presence of the end_stream timestamp as the indicator of completed client's stream.

This change can be reverted by setting the envoy.reloadable_features.use_filter_manager_state_for_downstream_end_stream runtime flag to false.

Risk Level: Medium, affects deferred request processing
Testing: Unit tests
Docs Changes: N/A
Release Notes: Yes
Platform Specific Features: N/A
Runtime guard: envoy.reloadable_features.use_filter_manager_state_for_downstream_end_stream
Fixes #35814

### CODE REVIEW

File: source/common/http/conn_manager_impl.cc [hunk 1]
```diff
@@ -1478,14 +1482,19 @@ void ConnectionManagerImpl::ActiveStream::decodeTrailers(RequestTrailerMapPtr&&
...
   }
-  maybeEndDecode(true);
+  maybeRecordLastByteReceived(true);
   if (!state_.deferred_to_next_io_iteration_) {
+    request_trailers_ = std::move(trailers);
     filter_manager_.decodeTrailers(*request_trailers_);
+  } else {
+    // Save trailers in a different variable since `request_trailers_` is available to the filter
+    // manager via `requestTrailers()` callback and makes filter manager see trailers prematurely
+    // when deferred request is processed.
```
Comment (RyanTheOptimist) [inline source/common/http/conn_manager_impl.cc:1496]: Nice comment! Thanks for adding this.

**Review by RyanTheOptimist — APPROVED** [review]
Wow, great catch! LGTM

**Review by KBaichoo — DISMISSED** [review]
Nice work!

File: test/integration/filters/stop_in_headers_continue_in_data_filter.cc [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+#include "source/extensions/filters/http/common/pass_through_filter.h"
+
+#include "test/extensions/filters/http/common/empty_http_filter_config.h"
+#include "test/integration/filters/common.h"
+
+namespace Envoy {
+
+// A test filter that stops iteration of headers request/response
+// then injects a body later.
```
[the remark this answers is not in the document]
Comment (yanavlasov) [reply test/integration/filters/stop_in_headers_continue_in_data_filter.cc]: Fixed

**Review by KBaichoo — APPROVED** [review]

**Pull request merged**